hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/FileGroupReaderBasedParquetFileFormat.scala

```diff
@@ -0,0 +1,255 @@
+                                              hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
+    val requiredSchemaWithMandatory = generateRequiredSchemaWithMandatory(requiredSchema, dataSchema)
+    val requiredSchemaSplits = requiredSchemaWithMandatory.fields.partition(f => HoodieRecord.HOODIE_META_COLUMNS_WITH_OPERATION.contains(f.name))
+    val requiredMeta = StructType(requiredSchemaSplits._1)
+    val requiredWithoutMeta = StructType(requiredSchemaSplits._2)
+    val (baseFileReader, preMergeBaseFileReader, _, _) = buildFileReaders(sparkSession,
+      dataSchema, partitionSchema, requiredSchema, filters, options, hadoopConf, requiredSchemaWithMandatory,
+      requiredWithoutMeta, requiredMeta)
+    val broadcastedHadoopConf = sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    (file: PartitionedFile) => {
+      file.partitionValues match {
+        case fileSliceMapping: PartitionFileSliceMapping =>
+          val filePath = sparkAdapter.getSparkPartitionedFileUtils.getPathFromPartitionedFile(file)
+          if (FSUtils.isLogFile(filePath)) {
+            // TODO: Use FileGroupReader here.
+            throw new NotImplementedError("Not support reading with only log files")
+          } else {
+            fileSliceMapping.getSlice(FSUtils.getFileId(filePath.getName)) match {
+              case Some(fileSlice) =>
+                val hoodieBaseFile = fileSlice.getBaseFile.get()
+                val bootstrapFileOpt = hoodieBaseFile.getBootstrapBaseFile
+                val partitionValues = fileSliceMapping.getInternalRow
+                val logFiles = getLogFilesFromSlice(fileSlice)
+                if (requiredSchemaWithMandatory.isEmpty) {
+                  val baseFile = createPartitionedFile(partitionValues, hoodieBaseFile.getHadoopPath, 0, hoodieBaseFile.getFileLen)
+                  // TODO: Use FileGroupReader here.
+                  baseFileReader(baseFile)
+                } else if (bootstrapFileOpt.isPresent) {
+                  // TODO: Use FileGroupReader here.
+                  throw new NotImplementedError("Not support reading bootstrap file")
+                } else {
+                  if (logFiles.isEmpty) {
+                    throw new IllegalStateException(
+                      "should not be here since file slice should not have been broadcasted "
+                        + "since it has no log or data files")
+                  }
+                  buildFileGroupIterator(
+                    preMergeBaseFileReader,
+                    partitionValues,
+                    hoodieBaseFile,
+                    logFiles,
+                    requiredSchemaWithMandatory,
+                    broadcastedHadoopConf.value.value,
+                    hoodieBaseFile.getFileLen,
+                    positionBasedWriteEnabled
+                  )
+                }
+              // TODO: Use FileGroupReader here.
```

Comment:
let's file followup JIRAs for different query types.

Reply:
Will do.